pkg/rtc/room.go

```diff
@@ -362,24 +371,34 @@ func (r *Room) UpdateSubscriptions(
 		for _, trackSid := range pt.TrackSids {
 			track := p.GetPublishedTrack(trackSid)
 			if track != nil {
-				tracks[trackSid] = track
+				trackInfos[trackSid] = struct {
+					publisher types.Participant
+					track     types.PublishedTrack
+				}{
+					publisher: p,
+					track:     track,
+				}
 			}
 		}
 	}
 
 	// handle subscription changes
-	for _, track := range tracks {
+	for trackSid, trackInfo := range trackInfos {
 		if subscribe {
-			if err := track.AddSubscriber(participant); err != nil {
+			if _, err := trackInfo.publisher.AddSubscriber(participant, types.AddSubscriberParams{TrackSids: []string{trackSid}}); err != nil {
 				return err
 			}
 		} else {
-			track.RemoveSubscriber(participant.ID())
+			trackInfo.track.RemoveSubscriber(participant.ID())
 		}
 	}
 	return nil
 }
 
+func (r *Room) UpdateSubscriptionPermissions(participant types.Participant, permissions *livekit.UpdateSubscriptionPermissions) error {
```

Comment:
New method in `Room` interface. Pretty much forwards it to the publisher participant with a participant resolver attached.